### PR TITLE
feat(core): support for using SPM (Swift Package Manager)

### PR DESCRIPTION
**Pull request by russellwheatley**
## Description

- Works with SPM and Cocoapods. To use SPM,  remove `Podfile` as well as run `pod deintegrate` in the `example/ios` directory. You will know it worked when you don't see `Running pod install...` in the logs when running `flutter run`.
- `Package.swift` does not have the same attributes that a podspec has. For example, you cannot set `version`, `summary`, `description`, `homepage`, `license` & `authors`.
- [Updated LIBRARY_VERSION and LIBRARY_NAME](https://github.com/firebase/flutterfire/pull/12786/files#diff-021dfe6df9eae0b2a7c718fd0df578115005602e6e476dece7e0b2a67139d7cbL108-L112) so the "@" symbol is in the code. I could not get the "@" character to escape in the `Package.swift` file causing build problem. I've also updated the [podspec file to match](https://github.com/firebase/flutterfire/pull/12786/files#diff-ddf09ba7a30b027748e9e65c63d6adbd9ab89c69d31a475d90c0e6810a34fd3fR40).
- Finally found a way to get firebase_core version and firebase-ios-sdk version dynamically in `Package.swift` file. Followed example of [brave here](https://github.com/brave/brave-core/blob/1da4e5adc5b070c8e81d8184189a6358303e20a7/ios/brave-ios/Package.swift#L602). This works when running in Xcode and via Flutter CLI.
- I've used [firebase app installations](https://github.com/firebase/flutterfire/pull/12786/files#diff-8ba779f9dea534ccdcb7bea8d78060e805ad75ef984ad7ce64d49c947ab76f16R84) as a product so we have access to firebase core, SPM doesn't ship firebase core as a standalone product.

## TODO
- [x] check how the other plugins work with the updates.
- [x] update macOS podspec in firebase_core.
- [x] write integration tests building with SPM.
- [x] update macOS `Package.swift` to also extract versions dynamically. Test on Xcode build and via Flutter CLI
- [x] Update integration test for macOS to check it also works via SPM. Extract into a shell script as it is becoming unmanageable.

 
## Related Issues

*Replace this paragraph with a list of issues related to this PR from the [issue database](https://github.com/flutter/flutter/issues). Indicate, which of these issues are resolved or fixed by this PR. Note that you'll have to prefix the issue numbers with flutter/flutter#.*

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`).
This will ensure a smooth and quick review process. Updating the `pubspec.yaml` and changelogs is not required.

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [ ] All existing and new tests are passing.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] The analyzer (`melos run analyze`) does not report any problems on my PR.
- [ ] I read and followed the [Flutter Style Guide].
- [ ] I signed the [CLA].
- [ ] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change.
- [ ] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/firebase/flutterfire/blob/master/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[CLA]: https://cla.developers.google.com/

### CODE REVIEW

File: packages/firebase_core/firebase_core/example/ios/Runner.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -9,6 +9,7 @@
 /* Begin PBXBuildFile section */
 		1498D2341E8E89220040F4C2 /* GeneratedPluginRegistrant.m in Sources */ = {isa = PBXBuildFile; fileRef = 1498D2331E8E89220040F4C2 /* GeneratedPluginRegistrant.m */; };
 		3B3967161E833CAA004F5970 /* AppFrameworkInfo.plist in Resources */ = {isa = PBXBuildFile; fileRef = 3B3967151E833CAA004F5970 /* AppFrameworkInfo.plist */; };
+		78A318202AECB46A00862997 /* FlutterGeneratedPluginSwiftPackage in Frameworks */ = {isa = PBXBuildFile; productRef = 78A3181F2AECB46A00862997 /* FlutterGeneratedPluginSwiftPackage */; };
```
Comment (loic-sharma) [inline packages/firebase_core/firebase_core/example/ios/Runner.xcodeproj/project.pbxproj:12]: FYI, checking in these project changes will make the example app require Flutter 3.24 or higher.

If this isn't acceptable, here are the instructions to [undo these changes](https://docs.flutter.dev/packages-and-plugins/swift-package-manager/for-app-developers#how-to-remove-swift-package-manager-integration).

File: packages/firebase_core/firebase_core/ios/firebase_core.podspec [hunk 2]
```diff
@@ -37,7 +37,7 @@ Pod::Spec.new do |s|
 
   s.static_framework = true
   s.pod_target_xcconfig = {
-    'GCC_PREPROCESSOR_DEFINITIONS' => "LIBRARY_VERSION=\\@\\\"#{library_version}\\\" LIBRARY_NAME=\\@\\\"flutter-fire-core\\\"",
+    'GCC_PREPROCESSOR_DEFINITIONS' => "LIBRARY_VERSION=\\\"#{library_version}\\\" LIBRARY_NAME=\\\"flutter-fire-core\\\"",
     'DEFINES_MODULE' => 'YES'
   }
-end
+end
```
Comment (loic-sharma) [inline packages/firebase_core/firebase_core/ios/firebase_core.podspec:43]:
```suggestion
end

```

File: packages/firebase_core/firebase_core/ios/firebase_core/Package.swift [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+}
+
+guard let firebase_sdk_version = Version(firebase_sdk_version_string) else {
+  fatalError("Invalid Firebase SDK version: \(firebase_sdk_version_string)")
+}
+
+let package = Package(
+  name: "firebase_core",
+  platforms: [
+    .iOS("11.0"),
```
Comment (loic-sharma) [inline packages/firebase_core/firebase_core/ios/firebase_core/Package.swift:85]: FYI, Flutter iOS requires 12.0+. You can bump this to iOS 12.0 if you'd like

**Comment by loic-sharma** [comment]
@russellwheatley Excellent work!

Did you follow Flutter's [Swift Package Manager migration guide](https://docs.flutter.dev/packages-and-plugins/swift-package-manager/for-plugin-authors) to create this pull request? What problems did you run into? Do you have any feedback on the guide and how we could improve it?

File: packages/firebase_core/firebase_core/example/ios/Runner.xcodeproj/project.pbxproj (hunk 1 above)
Comment (russellwheatley) [reply packages/firebase_core/firebase_core/example/ios/Runner.xcodeproj/project.pbxproj:12]: Thanks for the heads up! I think this is fine as 3.24 is the latest stable version 👍

**Comment by russellwheatley** [comment]
Hey @loic-sharma - thanks! Sorry for the late response. I actually used the instructions [on the pull request](https://github.com/flutter/flutter/pull/146256#issuecomment-2038565077) which seems like a precursor to the documentation you linked. It was actually a pretty good guide!

The only trouble I ran into was project specific such as being able to get the firebase-ios-sdk version from within Package.swift file, and also trying to parse preprocessor definitions 😄

File: packages/firebase_core/firebase_core/ios/firebase_core/Package.swift [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+}
+
+guard let firebase_sdk_version = Version(firebase_sdk_version_string) else {
+  fatalError("Invalid Firebase SDK version: \(firebase_sdk_version_string)")
+}
+
+let package = Package(
+  name: "firebase_core",
+  platforms: [
+    .iOS("12.0"),
```
Comment (Lyokone) [inline packages/firebase_core/firebase_core/ios/firebase_core/Package.swift:85]: should it be 13.0? https://firebase.google.com/support/release-notes/ios#version_1100_-_july_30_2024

Comment (russellwheatley) [reply packages/firebase_core/firebase_core/ios/firebase_core/Package.swift:85]: good catch 👍

**Review by exaby73 — APPROVED** [review]
